WPThemeReview/Sniffs/Templates/ReservedTemplateNameSniff.php

```diff
@@ -0,0 +1,156 @@
+	 * @link https://developer.wordpress.org/themes/template-files-section/partial-and-miscellaneous-template-files/#content-slug-php
+	 * @link https://wphierarchy.com/
+	 * @link https://en.wikipedia.org/wiki/Media_type#Naming
+	 *
+	 * @since 0.2.0
+	 *
+	 * @var array
+	 */
+	protected $reserved_file_name_prefixes = [
+		'author-'   => true,
+		'category-' => true,
+		'page-'     => true,
+		'tag-'      => true,
+	];
+
+	/**
+	 * Strip quotes surrounding an arbitrary string.
+	 *
+	 * Intended for use with the contents of a T_CONSTANT_ENCAPSED_STRING / T_DOUBLE_QUOTED_STRING.
+	 *
+	 * Copied from the WordPressCS\WordPress\Sniff abstract class.
+	 *
+	 * @since 0.2.0
+	 *
+	 * @param string $string The raw string.
+	 * @return string String without quotes around it.
+	 */
+	private function strip_quotes( $string ) {
+		return preg_replace( '`^([\'"])(.*)\1$`Ds', '$2', $string );
+	}
+
+	/**
+	 * Returns an array of tokens this test wants to listen for.
+	 *
+	 * @return array
+	 */
+	public function register() {
+		return array(
+			\T_OPEN_TAG,
+			\T_OPEN_TAG_WITH_ECHO,
+		);
+	}
+
+	/**
+	 * Processes this test, when one of its tokens is encountered.
+	 *
+	 * @param \PHP_CodeSniffer\Files\File $phpcsFile The PHP_CodeSniffer file where the
+	 *                                               token was found.
+	 * @param int                         $stackPtr  The position of the current token
+	 *                                               in the stack.
+	 *
+	 * @return void|int Optionally returns an integer stack pointer or void to continue
+	 *                  normal file processing.
+	 */
+	public function process( File $phpcsFile, $stackPtr ) {
+
+		// Usage of `strip_quotes` is to ensure `stdin_path` passed by IDEs does not include quotes.
+		$file = $this->strip_quotes( $phpcsFile->getFileName() );
+
+		$fileName = basename( $file );
+
+		if ( \defined( '\PHP_CODESNIFFER_IN_TESTS' ) ) {
+			$fileName = str_replace( '.inc', '.php', $fileName );
+		}
+
+		// Check if the current file has a prefix in the reserved list.
+		if ( ! $this->is_reserved_template_name_used( $fileName ) ) {
+			return ( $phpcsFile->numTokens + 1 );
+		}
+
+		$phpcsFile->addError(
+			self::ERROR_MSG,
+			0,
+			'ReservedTemplatePrefixFound',
+			array( $this->prefix, $this->slug )
+		);
+	}
+
+	/**
+	 * Checks if the given file is located in the $reserved_file_name_prefixes array.
+	 *
+	 * @param  string $file File name to check.
+	 * @return boolean
+	 */
+	private function is_reserved_template_name_used( $file ) {
+		$dash_position = strpos( $file, '-' );
+
+		if ( false === $dash_position || 0 === $dash_position ) {
+			// No dash or dash at start of filename, i.e. `-something.php`.
+			return false;
+		}
+
+		$partial = substr( $file, 0, $dash_position );
+		$slug    = substr( $file, $dash_position + 1, strlen( $file ) );
+
+		if ( empty( $partial ) ) {
+			return false;
+		}
+
+		if ( isset( $this->reserved_file_name_prefixes[ "$partial-" ] ) ) {
```

Comment:
That's because the `$partial` will return just the part without the dash, and I need to check if the found partial is in the `$reserved_file_name_prefixes` array which contains dashes, so this
`$this->reserved_file_name_prefixes[ "$partial-" ]` would not match anything because the array looks like this

```php
$reserved_file_name_prefixes = [
	'author-'   => true,
	'category-' => true,
	'page-'     => true,
	'tag-'      => true,
];
```

I could change the `$partial` to 

```php
substr( $file, 0, $dash_position + 1 );
```

And then use it to check, that could be better, right?

Reply:
Sorry, I think you missed my point - why do the array keys in the `$reserved_file_name_prefixes` all have a dash at the end ?

Reply:
Yes I changed that. When I read other comments I was like: why are there dashes in that array in the first place? 😄